chore: admit general PG booleans for configuration

jsilvela opened the pull request:
Allows any valid Postgres boolean for configuration in the webhooks.
Currently, only "off" and "on" are properly understood by the webhooks.

Closes #4252

jsilvela committed:
chore: admit general PG booleans for configuration

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela committed:
fix: add missing files

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela committed:
chore: make test clearer

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented:
/test limit=local feature_type=smoke

mnencia reviewed (changes requested):
I don't think it's correct to have two different functions (IsTrue and IsFalse) to parse a Postgres boolean.

We should have a single function `func ParsePostgresBoolean(in string) (bool, error)` and then manage the parse errors in the webhook.

Also, the PostgreSQL documentation says "Values can be written as on, off, true, false, yes, no, 1, 0 (all case-insensitive) or any unambiguous prefix of one of these.", and this implementation doesn't recognise "unambiguous prefixes".

jsilvela commented:
What is an "unambiguous prefix"? I've been looking for explanations or examples and found none.

jsilvela commented:
Because strictly speaking, an unambiguous prefix could be "fa", or "of". Is that something we really want?

mnencia commented:
> Because strictly speaking, an unambiguous prefix could be "fa", or "of". Is that something we really want?

Postgres does it https://doxygen.postgresql.org/bool_8c_source.html#l00036

jsilvela committed:
fix: binary parsing, archive_mode is not boolean

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented:
Realized, `archive_mode` is not a boolean but an enum. Leaving that one as was.

jsilvela committed:
chore: add test cases

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

gabriele-wolfox commented on api/v1/cluster_webhook.go:
```diff
@@ -2398,14 +2411,27 @@ func (r *Cluster) validatePgFailoverSlots() field.ErrorList {
+				field.Invalid(
+					field.NewPath("spec", "postgresql", "parameters", hotStandbyFeedbackKey),
+					hotStandbyFeedback,
+					fmt.Sprintf("invalid %s value. Must be a postgres boolean", hotStandbyFeedbackKey)))
```
I would suggest to change with 
`fmt.Sprintf("invalid `%s` value. Must be a postgres boolean", hotStandbyFeedbackKey)))`

So that the log format looks similar to the one at line 1208

mnencia replied:
Applied

gabriele-wolfox commented:
Beside my suggestion, PR looks good to me.

mnencia commented:
> Realized, `archive_mode` is not a boolean but an enum. Leaving that one as was.

You are right. Archive mode can only be `off`, `on` and `always`, so we can safely compare it to `off`.

mnencia commented on pkg/postgres/booleans.go:
```diff
@@ -0,0 +1,44 @@
+// See: https://www.postgresql.org/docs/current/config-setting.html
+// Boolean: Values can be written as on, off, true, false, yes, no, 1, 0 (all case-insensitive)
+// or any unambiguous prefix of one of these.
+func ParsePostgresBoolean(in string) (bool, error) {
+	trueValues := []string{"1", "on", "yes", "true", "y", "t", "ye", "tr", "tru"}
+	falseValues := []string{"0", "no", "off", "false", "n", "f", "of", "fa", "fal", "fals"}
+
+	sanitized := strings.TrimSpace(in)
+	sanitized = strings.ToLower(sanitized)
+
+	if slices.Contains(falseValues, sanitized) {
+		return false, nil
+	}
+	if slices.Contains(trueValues, sanitized) {
+		return true, nil
+	}
+	return false, fmt.Errorf("configuration value is not a postgres boolean: %s", in)
+}
```
What about this implementation?

```suggestion
func ParsePostgresBoolean(in string) (bool, error) {
	booleans := map[string]bool{
		"1":     true,
		"on":    true,
		"y":     true,
		"ye":    true,
		"yes":   true,
		"t":     true,
		"tr":    true,
		"tru":   true,
		"true":  true,
		"0":     false,
		"n":     false,
		"no":    false,
		"of":    false,
		"off":   false,
		"f":     false,
		"fa":    false,
		"fal":   false,
		"fals":  false,
		"false": false,
	}

	sanitized := strings.TrimSpace(in)
	sanitized = strings.ToLower(sanitized)

	if val, ok := booleans[sanitized]; ok {
		return val, nil
	}

	return false, fmt.Errorf("configuration value is not a postgres boolean: %s", in)
}
```

I find it easier to read.

gabriele-wolfox replied:
I prefer it, as you say it improve the readability.
+1

armru replied:
In my opinion, we should only accept two types and not try to do more, even if Postgres accepts them. This is because, in the end, they express only two values.
I propose that we accept only "true" "false" or "on" and "off" and we enable a strict validation on the webhook

mnencia replied:
Accepting only `on` and off` could surprise the user. Suppose you have set `hot_standby_feedback` to `yes` in your configuration, and it works because you don't have `pg_failover_slots` enabled.
You enable the pg_failover_slots extension, and the validation kicks in, rejecting a parameter that was valid before.

jsilvela replied:
I don't think this is an improvement. In the original, you can easily see the listing of "truthy" and "falsy" values.
With the map, you need to read the column on the right.
But, a `switch` statement would work well:
```suggestion
func ParsePostgresBoolean(in string) (bool, error) {
	sanitized := strings.TrimSpace(in)
	sanitized = strings.ToLower(sanitized)

	switch sanitized {
	case "1", "on", "yes", "true", "y", "t", "ye", "tr", "tru":
		return true, nil
	case "0", "no", "off", "false", "n", "f", "of", "fa", "fal", "fals":
		return false, nil
	default:
		return false, fmt.Errorf("configuration value is not a postgres boolean: %s", in)
	}
}
```

mnencia committed:
chore: review

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
chore: accepring gabriele-wolfox suggestion

Co-authored-by: Gabriele Quaresima <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Gabriele Quaresima <gabriele.quaresima@enterprisedb.com>
Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

jsilvela commented on pkg/postgres/booleans_test.go:
```diff
@@ -0,0 +1,49 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = DescribeTable("Moar PostgreSQL booleans parsing",
```
```suggestion
var _ = DescribeTable("Test PostgreSQL booleans parsing",
```

jsilvela committed:
Update pkg/postgres/booleans_test.go

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela commented on pkg/postgres/booleans_test.go:
```diff
@@ -0,0 +1,49 @@
+	. "github.com/onsi/gomega"
+)
+
+var _ = DescribeTable("Test PostgreSQL booleans parsing",
```
```suggestion
var _ = DescribeTable("Test parsing of PostgreSQL booleans",
```

jsilvela committed:
Update pkg/postgres/booleans_test.go

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

mnencia committed:
test: be esplicit about o being ambiguos

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

jsilvela committed:
chore: switch to switch

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

jsilvela committed:
chore: reorder

Signed-off-by: Jaime Silvela <jaime.silvela@enterprisedb.com>

mnencia committed:
chore: use correct version of tooling

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
chore: use the same order in for on and off

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>

mnencia committed:
review: PostgreSQL doesn't allow surrounding booleans with spaces

While it generally allow surrounding spaces when parsing booleans in SQL,
they are not permitted in the configuration file.

Adding a space before or after a value will cause the following log
entry on the reload and a failure on the restart:

{
  "level": "info",
  "ts": "2024-05-20T12:10:40Z",
  "logger": "postgres",
  "msg": "record",
  "logging_pod": "cluster-example-1",
  "record": {
    "log_time": "2024-05-20 12:10:40.498 UTC",
    "process_id": "1348",
    "session_id": "664b3da9.544",
    "session_line_num": "9",
    "session_start_time": "2024-05-20 12:10:17 UTC",
    "transaction_id": "0",
    "error_severity": "LOG",
    "sql_state_code": "22023",
    "message": "parameter \"wal_log_hints\" requires a Boolean value",
    "backend_type": "postmaster",
    "query_id": "0"
  }
}

Signed-off-by: Marco Nenciarini <marco.nenciarini@enterprisedb.com>
